weave/trace/weave_client.py

```diff
@@ -1970,34 +1975,190 @@ def _op_runs(self, op_def: Op) -> Sequence[Call]:
     def _ref_uri(self, name: str, version: str, path: str) -> str:
         return ObjectRef(self.entity, self.project, name, version).uri()
 
-    def flush(self) -> None:
+    def _send_file_create(self, req: FileCreateReq) -> Future[FileCreateRes]:
+        if self.future_executor_fastlane:
+            # If we have a separate upload worker pool, use it
+            return self.future_executor_fastlane.defer(self.server.file_create, req)
+        return self.future_executor.defer(self.server.file_create, req)
+
+    @property
+    def num_outstanding_jobs(self) -> int:
         """
-        An optional flushing method for the client.
-        Forces all background tasks to be processed, which ensures parallel processing
-        during main thread execution. Can improve performance when user code completes
-        before data has been uploaded to the server.
+        Returns the total number of pending jobs across all executors and the server.
+
+        This property can be used to check the progress of background tasks
+        without blocking the main thread.
+
+        Returns:
+            int: The total number of pending jobs
+        """
+        total = self.future_executor.num_outstanding_futures
+        if self.future_executor_fastlane:
+            total += self.future_executor_fastlane.num_outstanding_futures
+
+        # Add call batch uploads if available
+        if self._server_is_flushable:
+            total += self.server.call_processor.num_outstanding_jobs  # type: ignore
+        return total
+
+    def flush(
+        self,
+        use_progress_bar: bool = True,
+        callback: Callable[[WeaveFlushStatus], None] | None = None,
+    ) -> None:
+        """
+        Flushes all background tasks to ensure they are processed.
+
+        This method blocks until all currently enqueued jobs are processed,
+        displaying a progress bar to show the status of the pending tasks.
+        It ensures parallel processing during main thread execution and can
+        improve performance when user code completes before data has been
+        uploaded to the server.
+
+        Args:
+            use_progress_bar: Whether to display a progress bar during flush.
+                              Set to False for environments where a progress bar
+                              would not render well (e.g., CI environments).
+            callback: Optional callback function that receives status updates.
+                      Overrides use_progress_bar.
+        """
+        if use_progress_bar and callback is None:
+            callback = create_progress_bar_callback()
+        if callback is None:
+            self._flush()
+        else:
+            self._flush_with_callback(callback=callback)
+
+    def _flush_with_callback(
+        self,
+        callback: Callable[[WeaveFlushStatus], None] | None = None,
```

Comment:
callback should never be non here - would recommend removing that and simplifying the code a bit

Reply:
yes true